rayon-core/src/registry.rs

```diff
@@ -520,6 +536,19 @@ unsafe fn main_loop(worker: Worker<JobRef>, registry: Arc<Registry>, index: usiz
 
     // Normal termination, do not abort.
     mem::forget(abort_guard);
+
+    // Inform a user callback that we exited a thread.
```

Comment:
Actually, the answer here is (except in extreme conditions) yes. Generally, whenever user code panics, we catch the unwinding and then resume the unwinding over on one of the user's threads. Panics should never unwind up into this code we're looking at there.

The only reason that a user-induced panic could cause us not to reach this code is when it occurs due to `Drop` -- e.g., if dropping the user's closure triggers a panic, we won't catch and re-route that. In that case, we'll hit the `abort_guard` above and abort the entire process, basically. 

Similarly, in futures, there are some random cases where panics -- while caught -- cannot be clearly propagated anywhere, in which case we invoke the panic handler, which (typically) aborts. In the case where an abort occurred, obviously we won't execute this code.